assembly/src/parsers/ast/mod.rs

```diff
@@ -0,0 +1,658 @@
+    let program_start = tokens.pos();
+    // consume the 'begin' token
+    let header = tokens.read().expect("missing program header");
+    header.validate_begin()?;
+    tokens.advance();
+
+    // make sure there is something to be read
+    let start_pos = tokens.pos();
+    if tokens.eof() {
+        return Err(AssemblyError::unexpected_eof(start_pos));
+    }
+
+    // parse the sequence of nodes and add each node to the list
+    let mut end_of_nodes = false;
+    let beginning_node_count = program.body.len();
+    while !end_of_nodes {
+        let node_count = program.body.len();
+        parse_token(context, tokens, &program.procedures, &mut program.body)?;
+        end_of_nodes = program.body.len() == node_count;
+    }
+
+    // make sure at least one block has been read
+    if program.body.len() == beginning_node_count {
+        let start_op = tokens.read_at(start_pos).expect("no start token");
+        return Err(AssemblyError::empty_block(start_op));
+    }
+
+    // consume the 'end' token
+    match tokens.read() {
+        None => Err(AssemblyError::unmatched_begin(
+            tokens.read_at(program_start).expect("no begin token"),
+        )),
+        Some(token) => match token.parts()[0] {
+            Token::END => token.validate_end(),
+            Token::ELSE => Err(AssemblyError::dangling_else(token)),
+            _ => Err(AssemblyError::unmatched_begin(
+                tokens.read_at(program_start).expect("no begin token"),
+            )),
+        },
+    }?;
+    tokens.advance();
+
+    Ok(())
+}
+
+// Parse procedure from token stream.
+fn parse_procedure(
+    context: &ParserContext,
+    tokens: &mut TokenStream,
+    procedures: &ProcMap,
+) -> Result<types::Procedure, AssemblyError> {
+    let proc_start = tokens.pos();
+
+    // read procedure name and consume the procedure header token
+    let header = tokens.read().expect("missing procedure header");
+    let (label, num_locals, is_export) = header.parse_proc()?;
+    tokens.advance();
+
+    let mut body = Vec::<Node>::new();
+
+    // parse procedure body
+    parse_token(context, tokens, procedures, &mut body)?;
+
+    // consume the 'end' token
+    match tokens.read() {
+        None => Err(AssemblyError::unmatched_proc(
+            tokens.read_at(proc_start).expect("no proc token"),
+        )),
+        Some(token) => match token.parts()[0] {
+            Token::END => token.validate_end(),
+            _ => Err(AssemblyError::unmatched_proc(
+                tokens.read_at(proc_start).expect("no proc token"),
+            )),
+        },
+    }?;
+    tokens.advance();
+
+    // build and return the procedure
+    Ok(types::Procedure {
+        name: label,
+        num_locals,
+        is_export,
+        body,
+        index: 0,
+    })
+}
+
+// parse if and else tokens into AST nodes.
+fn parse_if(
+    context: &ParserContext,
+    tokens: &mut TokenStream,
+    procedures: &ProcMap,
+    nodes: &mut Vec<Node>,
```

Comment:
Is there a reason to take `nodes` as a mutable parameter here instead of returning `Node`?

And similar comment for `parse_while()`, `parse_repeat()` etc.

Reply:
Trying to remove allocations, but can also juts let it return nodes if you want

Reply:
I don't think we need to return a vector of nodes here - we just return a single node, right?

Reply:
Yes that's true, though each node holds a vector of nodes, so it does gets copied out depending on how deep we go. For now I won't worry about it given your suggestion

Reply:
Right - but there are only two functions which needs to take a mutable vector as a parameter: `parse_token` and `parse_body` - everything else can just return a node. Also, once the suggested refactoring is done, we should be able to get rid of `parse_body` (and we can rename `parse_token` into `parse_body`). This way, there will be only one function which takes a mutable vector.

Reply:
fixed